flink-ml-lib/src/main/java/org/apache/flink/ml/feature/vectorassembler/VectorAssembler.java

```diff
@@ -74,64 +81,96 @@ public Table[] transform(Table... inputs) {
+            keepInvalid = handleInvalid.equals(HasHandleInvalid.KEEP_INVALID);
         }
 
         @Override
         public void flatMap(Row value, Collector<Row> out) {
             int nnz = 0;
-            int vectorSize = 0;
             try {
-                for (String inputCol : inputCols) {
-                    Object object = value.getField(inputCol);
-                    Preconditions.checkNotNull(object, "Input column value should not be null.");
+                for (int i = 0; i < inputCols.length; ++i) {
+                    Object object = value.getField(inputCols[i]);
+                    if (object == null) {
+                        if (keepInvalid) {
+                            if (inputSizes[i] > 1) {
+                                DenseVector tmpVec = new DenseVector(inputSizes[i]);
+                                for (int j = 0; j < inputSizes[i]; ++j) {
+                                    tmpVec.values[j] = Double.NaN;
+                                }
+                                object = tmpVec;
+                            } else {
+                                object = Double.NaN;
+                            }
+                            value.setField(inputCols[i], object);
+                        } else {
+                            throw new RuntimeException(
+                                    "Input column value is null. Please check the input data or using handleInvalid = 'keep'.");
+                        }
+                    }
                     if (object instanceof Number) {
+                        if (Double.isNaN(((Number) object).doubleValue()) && !keepInvalid) {
+                            throw new RuntimeException(
+                                    "Encountered NaN while assembling a row with handleInvalid = 'error'. Consider "
+                                            + "removing NaNs from dataset or using handleInvalid = 'keep' or 'skip'.");
+                        }
+                        checkVectorAndNumberSize(inputSizes[i], 1);
                         nnz += 1;
-                        vectorSize += 1;
                     } else if (object instanceof SparseVector) {
+                        int localSize = ((SparseVector) object).size();
+                        checkVectorAndNumberSize(inputSizes[i], localSize);
                         nnz += ((SparseVector) object).indices.length;
-                        vectorSize += ((SparseVector) object).size();
                     } else if (object instanceof DenseVector) {
+                        int localSize = ((DenseVector) object).size();
+                        checkVectorAndNumberSize(inputSizes[i], localSize);
                         nnz += ((DenseVector) object).size();
-                        vectorSize += ((DenseVector) object).size();
                     } else {
                         throw new IllegalArgumentException(
                                 "Input type has not been supported yet.");
```

Comment:
Let's add descriptions about what is the type of input value that has not been supported yet.